File: app/src/main/java/com/gnardini/tvshowcontroller/networking/RetrofitServices.java
package com.gnardini.tvshowcontroller.networking;

import java.util.HashMap;
import java.util.Map;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitServices {

    private final Map<Class, Object> services;
    private final HostManager hostManager;

    public RetrofitServices(HostManager hostManager) {
        this.services = new HashMap<>();
        this.hostManager = hostManager;
    }

    public <ServiceType> ServiceType createService(Class<ServiceType> clazz) {
//        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
//        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder()
//                .addInterceptor(interceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(hostManager.getHost())
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        return retrofit.create(clazz);
    }
}

File: app/src/main/java/com/gnardini/tvshowcontroller/networking/RetrofitServices.kt
package com.gnardini.tvshowcontroller.networking

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.*

class RetrofitServices(private val hostManager: HostManager) {

    private val services = HashMap<Class<*>, Any>()
    private var retrofit = createRetrofit()

    fun <ServiceType> getService(clazz: Class<ServiceType>): ServiceType {
        if (!services.containsKey(clazz)) {
            val service: ServiceType = retrofit.create(clazz)
            services.put(clazz, service as Any)
        }
        return services.get(clazz) as ServiceType
    }

    fun refresh() {
        createRetrofit()
    }

    private fun createRetrofit(): Retrofit {
        retrofit = Retrofit.Builder()
                .baseUrl(hostManager.getHost())
                .client(createClient())
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        return retrofit
    }

    private fun createClient(): OkHttpClient {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        return OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build()
    }
}
